pkg/scheduler/core/division_algorithm.go

```diff
@@ -20,6 +20,82 @@ func (a TargetClustersList) Len() int           { return len(a) }
 func (a TargetClustersList) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
 func (a TargetClustersList) Less(i, j int) bool { return a[i].Replicas > a[j].Replicas }
 
+type dispenser struct {
+	numReplicas int32
+	result      []workv1alpha2.TargetCluster
+}
+
+func newDispenser(numReplicas int32, init []workv1alpha2.TargetCluster) *dispenser {
+	cp := make([]workv1alpha2.TargetCluster, len(init))
+	copy(cp, init)
+	return &dispenser{numReplicas: numReplicas, result: cp}
+}
+
+func (a *dispenser) done() bool {
+	return a.numReplicas == 0 && len(a.result) != 0
+}
+
+func (a *dispenser) takeByWeight(w helper.ClusterWeightInfoList) {
+	if a.done() {
+		return
+	}
+	sum := w.GetWeightSum()
+	if sum == 0 {
+		return
+	}
+
+	sort.Sort(w)
+
+	result := make([]workv1alpha2.TargetCluster, 0, w.Len())
+	remain := a.numReplicas
+	for _, info := range w {
+		replicas := int32(info.Weight * int64(a.numReplicas) / sum)
+		result = append(result, workv1alpha2.TargetCluster{
+			Name:     info.ClusterName,
+			Replicas: replicas,
+		})
+		remain -= replicas
+	}
+	// TODO(Garrybest): take rest replicas by fraction part
```

Comment:
I think this doesnt need more works, it looks fine

Reply:
Well, this assignment may be a little unfair in some scenario. Let me give a example here.

1. We have 3 clusters which have a weight of [A:1, B:2, C:3].
2. The desired replicas is 2.

According to this alg, we may have a result of [A:1, B:1], which is not fair because C got a higher weight.

Reply:
I did the test, looks like we need to think more about this:
```go
		{
			name:        "tmp",
			numReplicas: 2,
			result: []workv1alpha2.TargetCluster{
				{Name: "A", Replicas: 0},
				{Name: "B", Replicas: 0},
				{Name: "C", Replicas: 0},
			},
			weightList: []utilhelper.ClusterWeightInfo{
				{ClusterName: "A", Weight: 1},
				{ClusterName: "B", Weight: 2},
				{ClusterName: "C", Weight: 3},
			},
			desired: []workv1alpha2.TargetCluster{
				{Name: "A", Replicas: 0},
				{Name: "B", Replicas: 1},
				{Name: "C", Replicas: 1},
			},
			done: true,
		},
```
The result:
![image](https://user-images.githubusercontent.com/19745536/204083007-fdcfc4c4-0ca1-409a-b253-190d581f7034.png)

Reply:
Sorry, this case is not proper because now we have sort the cluster by weight in descent order(I didn't notice before). Consider we have a weight of [A:5, B:4, C:1] and desired replicas is 9, the result is [A:5, B:4, C:0]. 

Actually, the weighted result is [A:4.5, B:3.6, C:0.9], I think [A:4, B:4, C:1] should be fair because 0.9 > 0.6 > 0.5, i.e., the rest replicas of C is closer to 1. This case demonstrates the TODO list, I think regarding the fraction part of weighted result and sorting them by descent order could be more accurate mathematically.🤣

Reply:
Completely agree.